solr/core/src/java/org/apache/solr/core/TransientSolrCoreCacheDefault.java

```diff
@@ -89,6 +86,25 @@ public TransientSolrCoreCacheDefault(CoreContainer coreContainer) {
     transientDescriptors = new LinkedHashMap<>(initialCapacity);
   }
 
+  private void onEvict(SolrCore core) {
+    if (coreContainer.hasPendingCoreOps(core.getName())) {
+      if (log.isInfoEnabled()) {
+        log.info("NOT evicting transient core [{}]; it's loading or something else", core.getName());
+      }
+      transientCores.put(core.getName(), core); // put back
```

Comment:
In a race scenario with the removalListener: getCore will call `core = solrCores.waitAddPendingCoreOps(name);` which will return the core that was put back (un-evicted).  It will see it and return it without creating a new one.  I don't think there will be errors.